### PR TITLE
Updated PyTorch Lightning, added optional inference/validation loop, added custom loss functions

### PR DESCRIPTION
**Pull request by fredwang25**
- Updated PyTorch Lightning to 1.6.5.
- added the validation loop where we can (optionally) specify validation data, so we can compute validation loss, NRMSE, and SSIM and plot on Tensorboard. Specify with `data_val_file` and `num_val_data_sets` in the config (training config has been renamed `data_train_file` and `num_train_data_sets`).
- added support for L1, L2, and unweighted SSIM loss functions, specify with `loss_function` in the config.
- can now (optionally) specify datasets for inference, and it will automatically reconstruct the specified datasets after training is finished. Specify with `data_inference_file`, `data_inference_output`, and `num_inference_data_sets` in the config.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -66,16 +66,16 @@ def main_train(args, gpu_ids=None):
...
-        distributed_backend = None # FIXME should this also be ddp?
+        accelerator = None
     elif args.hyperopt:
         gpus = 1
-        distributed_backend = None
+        accelerator = None
     else:
         gpus = gpu_ids
-        distributed_backend = 'ddp'
+        accelerator = "ddp"
```
Comment (jtamir) [inline main.py]: can you use single quotes here?

File: deepinpy/recons/recon.py [hunk 2]
```diff
@@ -18,10 +18,20 @@
...
+def calc_nrmse_real(gt, pred):
     resid = pred - gt
     return (torch.real(opt.zdot_single_batch(resid)) / torch.real(opt.zdot_single_batch(gt))).sqrt().mean()
 
+def calc_nrmse_abs(gt, pred):
+    temp = torch.sqrt(torch.mean(torch.square(torch.abs(gt) - torch.abs(pred))))
+    rr = torch.max(torch.abs(gt)) - torch.min(torch.abs(gt))  # range
+    return torch.div(temp, rr)
+
+def calc_nrmse(gt, pred, nrmse_only_real):
```
Comment (jtamir) [inline deepinpy/recons/recon.py]: can you add a default value, i.e. `nrmse_only_real=True` ?
Also, perhaps it's clearer to use the variable `absval=False` and switch the logic

File: deepinpy/recons/recon.py [hunk 3]
```diff
@@ -18,10 +18,20 @@
 from torch.optim import lr_scheduler 
 
 @torch.jit.script
-def calc_nrmse(gt, pred):
+def calc_nrmse_real(gt, pred):
     resid = pred - gt
     return (torch.real(opt.zdot_single_batch(resid)) / torch.real(opt.zdot_single_batch(gt))).sqrt().mean()
 
+def calc_nrmse_abs(gt, pred):
```
Comment (jtamir) [inline deepinpy/recons/recon.py]: I understand your intention, but this is not the same as the NRMSE of the absolute value between x and y. Perhaps you should give this function a different name to reflect that.

File: deepinpy/recons/recon.py [hunk 4]
```diff
@@ -56,13 +67,24 @@ def _init_hparams(self, hparams):
...
+        if loss_type == "L1":
+            resid = torch.abs(pred - gt)
+            return torch.mean(resid)
+        elif loss_type == "L2":
+            resid = torch.abs(torch.square(pred - gt))
+            return torch.mean(resid)
+        elif loss_type == "SSIM":
+            return utils.ssim_loss(pred, gt)
+        elif loss_type is None:
+            return torch.mean(torch.real(opt.zdot_single_batch(pred - gt)))
```
Comment (jtamir) [inline deepinpy/recons/recon.py:63]: this should be equivalent to your L2 loss, can you please check? And if so, can you go with the one that gives better performance and get rid of the other?

File: deepinpy/recons/recon.py (hunk 3 above)
Comment (jtamir) [reply deepinpy/recons/recon.py]: In addition, I believe if the hyperparam `abs_loss` is set, then this will already calculate the NRMSE of the absolute value. More generally, it will calculate the loss w.r.t. to the absolute value of `gt` and the absolute value of `pred`

File: deepinpy/recons/recon.py [hunk 5]
```diff
@@ -115,6 +137,61 @@ def log_metadata(self, log_dict, key, fun=None):
...
+        Args:
+            batch (tuple): Should hold the indices of data and the corresponding data, in said order.
+            batch_idx (None): Currently unimplemented.
+
+        Returns:
+            A list of validation loss, NRMSE, and SSIM.
+        """
+        if self.hparams.data_val_file:
+            idx, data = batch
+            imgs = data["imgs"]
```
Comment (jtamir) [inline deepinpy/recons/recon.py]: please use single quotes for all string variables

File: deepinpy/utils/utils.py [hunk 6]
```diff
@@ -208,3 +209,88 @@ def ifft2(x, axes=(-2, -1)):
...
+
+    g = torch.exp(
+        -(
+            (torch.from_numpy(x_data) ** 2 + torch.from_numpy(y_data) ** 2)
+            / (2.0 * torch.tensor(sigma) ** 2)
+        )
+    )
+    return g / torch.sum(g)
+
+def ssim_loss(y_pred, y_true):
```
Comment (jtamir) [inline deepinpy/utils/utils.py:242]: Is there a reason you are using this over the SSIM loss function defined using `skimage` (https://github.com/facebookresearch/fastMRI/blob/main/fastmri/evaluate.py#L40) ?

File: main.py [hunk 7]
```diff
@@ -66,19 +68,33 @@ def main_train(args, gpu_ids=None):
...
 
 
-    trainer = Trainer(max_epochs=args.num_epochs, gpus=gpus, logger=tt_logger, checkpoint_callback=checkpoint_callback, distributed_backend=distributed_backend, accumulate_grad_batches=args.num_accumulate, progress_bar_refresh_rate=1, gradient_clip_val=args.clip_grads)
+    trainer = Trainer(max_epochs=args.num_epochs, gpus=gpus, logger=tt_logger, checkpoint_callback=checkpoint_callback, accelerator=accelerator, accumulate_grad_batches=args.num_accumulate, progress_bar_refresh_rate=1, gradient_clip_val=args.clip_grads)
 
     trainer.fit(M)
 
+    # Inference on some test dataset, if specified.
```
Comment (jtamir) [inline main.py:84]: In general my preference is for main.py to have no reference to the MultiChannelMRIDataset, as in principle this package could be used for non-MRI problems. Is there a way you could refactor this to be the case? Perhaps it is wishful thinking to maintain that separation.

**Comment by jtamir** [comment]
Thanks for your contribution. Could you please review my individual comments and let me know your thoughts? In addition, could you please update the PR message to be more instructive?
Thanks!

File: deepinpy/utils/utils.py (hunk 6 above)
Comment (fredwang25) [reply deepinpy/utils/utils.py:242]: This one natively uses torch tensors and we can also change the weighting of alpha, beta, and gamma in the SSIM calculation. (it's also from GE research, so we figured it's pretty reputable)

File: deepinpy/recons/recon.py (hunk 4 above)
Comment (fredwang25) [reply deepinpy/recons/recon.py:63]: In your `zdot_batch` function, the output tensor is reshaped into shape `(batch, -1)` and summed along the non-0 axis. So for multidimensional tensors, say of size `(x, y)`, when calculating the mean of the residuals we would only be dividing by x, not x*y. So if x << y, then the loss of `(x, y)` would be a lot less than the loss of the identical transposed `(y, x)` tensor. This explains why I saw they were not equivalent, is this intentional?

(this also applies to NRMSE calculation)

File: deepinpy/recons/recon.py (hunk 3 above)
Comment (fredwang25) [reply deepinpy/recons/recon.py]: I don't think `abs_loss` currently has any effect on the NRMSE, but I can just use that logic to simply things.

File: deepinpy/utils/utils.py (hunk 6 above)
Comment (jtamir) [reply deepinpy/utils/utils.py:242]: sounds good!

File: deepinpy/recons/recon.py (hunk 4 above)
Comment (jtamir) [reply deepinpy/recons/recon.py:63]: All tensors should have a batch dimension as the first dimension, for example (`from docs/getting_started.md`), `[Ntraining, N1, N2, ..., NT, X, Y, Z]: np.complex`.  If your data is a single sample, you should be setting the dimensions to `[1, N1, N2, ...]`. Let me know if that isn't clear, and if it is, let me know if you still have an issue.

File: deepinpy/recons/recon.py (hunk 3 above)
Comment (jtamir) [reply deepinpy/recons/recon.py]: Sounds good. I understand this is to log statistics and not for actual loss functions. My issue is that you are naming it `nrmse_abs` but your implementation is normalizing by the max/min, which is not the definition of NRMSE.

Comment (fredwang25) [reply deepinpy/recons/recon.py]: Edit: I'll just keep your implementation of the NRMSE. I didn't realize you were already normalizing by the norm of the ground truth

File: deepinpy/recons/recon.py (hunk 4 above)
Comment (fredwang25) [reply deepinpy/recons/recon.py:63]: Oh, I think I get it now. I was averaging the loss over each individual pixel, and it seems like you're averaging the loss over the entire image, which is why our L2 implementations were returning different values. For consistency, I can make the rest of my loss functions also average over the whole image.

**Comment by fredwang25** [comment]
Should be good to go now.

**Comment by jtamir** [comment]
Will look later, but definition of nrmse is norm(x-xhat)/norm(x), where norm is l2-norm. Just call it something else (since you divide by max - min)

Jon
www.jtsense.com
________________________________

**Comment by jtamir** [comment]
Looks great! As I still haven't put in automatic testing, I will test this tomorrow with `python -m unittest` to make sure everything is Kosher. Have you already tried to do that?

**Comment by fredwang25** [comment]
Sorry for the late response, just updated the unit tests with the new arguments, and everything passes.
